File: matlab/+awr2944/dopplerFFT.m
function [dopplerSpectrum, velAxis] = dopplerFFT(rangeCube, profile, varargin)
%DOPPLERFFT Doppler FFT on complex range spectrum.
%
%   [dopplerSpectrum, velAxis] = awr2944.dopplerFFT(rangeCube, profile)
%
%   rangeCube : complex double [rangeBins, rx, chirp, frame]
%   profile   : struct from smokeV1Profile()
%
%   Optional Parameters:
%       'Window'    — 'rectangular', 'hann', 'hamming' (default: 'hann')
%       'NFFT'      — Doppler FFT length (default: profile.chirps_per_frame)
%       'Clutter'   — 'none', 'slow_time_mean' (default: 'slow_time_mean')
%       'Normalize' — coherent gain normalization (default: true)
%
%   Returns:
%       dopplerSpectrum : complex double [rangeBins, rx, doppler, frame]
%       velAxis         : double [doppler, 1] — velocity in m/s (centered)

    p = inputParser;
    addParameter(p, 'Window', 'hann', @ischar);
    addParameter(p, 'NFFT', profile.chirps_per_frame, @isnumeric);
    addParameter(p, 'Clutter', 'slow_time_mean', @ischar);
    addParameter(p, 'Normalize', true, @islogical);
    parse(p, varargin{:});

    windowType = p.Results.Window;
    nfft       = p.Results.NFFT;
    clutterMode = p.Results.Clutter;
    normalize  = p.Results.Normalize;

    [nRange, nRx, nChirp, nFrame] = size(rangeCube);

    cube = rangeCube;

    % 1. Clutter removal
    switch lower(clutterMode)
        case 'none'
            % no-op
        case 'slow_time_mean'
            % Subtract mean across chirps (dim 3) per (range, rx, frame)
            cube = cube - mean(cube, 3);
        otherwise
            error('awr2944:dopplerFFT', 'Unknown clutter mode: %s', clutterMode);
    end

    % 2. Doppler window along chirp axis (dim 3)
    switch lower(windowType)
        case 'rectangular'
            win = ones(nChirp, 1);
        case 'hann'
            win = hann(nChirp, 'periodic');
        case 'hamming'
            win = hamming(nChirp, 'periodic');
        otherwise
            error('awr2944:dopplerFFT', 'Unknown window: %s', windowType);
    end

    if normalize
        win = win / mean(win);
    end

    % Reshape for broadcasting: [1, 1, nChirp, 1]
    win4d = reshape(win, 1, 1, [], 1);
    cube = cube .* win4d;

    % 3. FFT along chirp axis (dim 3)
    dopplerSpectrum = fft(cube, nfft, 3);

    % 4. fftshift along Doppler axis
    dopplerSpectrum = fftshift(dopplerSpectrum, 3);

    % 5. Velocity axis
    c = 299792458.0;
    lambda = c / profile.start_frequency_hz;
    Tc = profile.chirp_period_s;
    k = (0:nfft-1)' - nfft/2;
    fd = k / (nfft * Tc);
    velAxis = lambda * fd / 2;
end

File: src/awr2944_dca/dsp/doppler_fft.py
"""Doppler FFT and range-Doppler processing.

Operates on the complex range FFT cube [frame, chirp, rx, range_bin].
Each frame is processed independently — chirps are never concatenated
across frame boundaries.

Pipeline per frame:
    1. Optional clutter removal (slow-time mean subtraction along chirp axis)
    2. Apply Doppler window along chirp axis
    3. FFT along chirp axis
    4. fftshift to center zero-Doppler
    5. Noncoherent RX power combination (optional)

Output shape: [frame, doppler, rx, range_bin]
"""

from __future__ import annotations

import numpy as np

from awr2944_dca.dsp.config import (
    ClutterRemovalMode,
    DopplerProcessingConfig,
    RadarProfile,
    RxCombination,
)
from awr2944_dca.dsp.clutter import remove_clutter
from awr2944_dca.dsp.types import DopplerResult
from awr2944_dca.dsp.windows import get_window


def compute_doppler_fft(
    range_cube: np.ndarray,
    profile: RadarProfile,
    config: DopplerProcessingConfig | None = None,
    *,
    kaiser_beta: float = 6.0,
) -> DopplerResult:
    """Compute the Doppler FFT on a complex range cube.

    Parameters
    ----------
    range_cube : ndarray, complex64
        Shape [frame, chirp, rx, range_bin] — output of range FFT.
    profile : RadarProfile
    config : DopplerProcessingConfig, optional
    kaiser_beta : float
        Beta for Kaiser window.

    Returns
    -------
    DopplerResult
        Contains complex cube [frame, doppler, rx, range_bin],
        power_db, and rx_combined_power arrays.
    """
    if config is None:
        config = DopplerProcessingConfig()

    cube = range_cube.copy()
    num_chirps = cube.shape[1]

    # 1. Clutter removal
    cube = remove_clutter(cube, config.clutter_removal, config.zero_doppler_notch_bins)

    # 2. Doppler window along chirp axis
    win, cg = get_window(
        config.window,
        num_chirps,
        normalize=config.normalize_window,
        kaiser_beta=kaiser_beta,
    )
    # Reshape for broadcasting: [1, chirps, 1, 1]
    win_4d = win.reshape(1, -1, 1, 1)
    cube = cube * win_4d

    # 3. FFT along chirp axis (axis=1)
    complex_cube = np.fft.fft(cube, n=config.nfft, axis=1).astype(np.complex64)

    # 4. fftshift along Doppler axis
    complex_cube = np.fft.fftshift(complex_cube, axes=1)

    # 5. Power in dB
    magnitude = np.abs(complex_cube)
    eps = np.finfo(np.float32).tiny
    power_db = (20.0 * np.log10(np.maximum(magnitude, eps))).astype(np.float32)

    # 6. RX combination
    if config.rx_combination == RxCombination.NONCOHERENT_POWER:
        # Sum power across RX channels: |X|^2 summed over rx axis
        rx_power = np.sum(magnitude ** 2, axis=2)  # [frame, doppler, range_bin]
        rx_combined_db = (10.0 * np.log10(np.maximum(rx_power, eps))).astype(np.float32)
    else:
        # No combination — pick RX0 as placeholder
        rx_combined_db = power_db[:, :, 0, :]

    return DopplerResult(
        complex_cube=complex_cube,
        power_db=power_db,
        rx_combined_power=rx_combined_db,
        config_dict={
            "profile": profile.to_dict(),
            "doppler_config": config.to_dict(),
            "kaiser_beta": kaiser_beta,
            "coherent_gain": cg,
            "num_input_chirps": num_chirps,
            "nfft": config.nfft,
        },
    )
